File: 4-envelope-encryption-and-kms/2-csharp/EnvelopeService.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using VaultSharp;
using VaultSharp.V1.AuthMethods.Token;
using VaultSharp.V1.SecretsEngines.Transit;

namespace EnvelopeApi;

public record Envelope(string WrappedDek, string Iv, string Ciphertext);

/// <summary>
/// Two-tier envelope encryption:
///   data ──AES-256-GCM──▶ encrypted by a per-record DEK (generated locally)
///   DEK  ──Transit wrap──▶ encrypted by the KEK (lives only in Vault)
/// Bulk data never touches Vault; only the 32-byte DEK is wrapped/unwrapped.
/// </summary>
public class EnvelopeService
{
    private readonly IVaultClient _vault;
    private readonly HttpClient _http;
    private readonly string _addr;
    private readonly string _token;
    private readonly string _kek;

    public EnvelopeService(string addr, string token, string kek)
    {
        _addr = addr;
        _token = token;
        _kek = kek;
        _vault = new VaultClient(new VaultClientSettings(addr, new TokenAuthMethodInfo(token)));
        _http = new HttpClient { BaseAddress = new Uri(addr) };
        _http.DefaultRequestHeaders.Add("X-Vault-Token", token);
    }

    /// <summary>Enable Transit and create the KEK (idempotent) at boot.</summary>
    public async Task EnsureKekAsync()
    {
        await PostRawAsync($"/v1/sys/mounts/transit", new { type = "transit" });
        await PostRawAsync($"/v1/transit/keys/{_kek}", new { type = "aes256-gcm96" });
    }

    public async Task<Envelope> EncryptAsync(string plaintext)
    {
        // 1. Fresh 256-bit DEK + 96-bit GCM nonce for THIS record only.
        var dek = RandomNumberGenerator.GetBytes(32);
        var iv = RandomNumberGenerator.GetBytes(12);

        // 2. Encrypt the bulk data locally with AES-256-GCM — Vault never sees the plaintext.
        var plainBytes = Encoding.UTF8.GetBytes(plaintext);
        var cipherBytes = new byte[plainBytes.Length];
        var tag = new byte[16];
        using (var aes = new AesGcm(dek, 16))
        {
            aes.Encrypt(iv, plainBytes, cipherBytes, tag);
        }
        var sealedBytes = cipherBytes.Concat(tag).ToArray(); // ciphertext||authTag

        // 3. Wrap the DEK with the KEK inside Vault.
        var enc = await _vault.V1.Secrets.Transit.EncryptAsync(_kek, new EncryptRequestOptions
        {
            Base64EncodedPlainText = Convert.ToBase64String(dek)
        });

        return new Envelope(enc.Data.CipherText, Convert.ToBase64String(iv), Convert.ToBase64String(sealedBytes));
    }

    public async Task<string> DecryptAsync(string wrappedDek, string ivB64, string ctB64)
    {
        // 1. Ask Vault to unwrap the DEK (works across KEK versions after rotation).
        var dec = await _vault.V1.Secrets.Transit.DecryptAsync(_kek, new DecryptRequestOptions
        {
            CipherText = wrappedDek
        });
        var dek = Convert.FromBase64String(dec.Data.Base64EncodedPlainText);
        var iv = Convert.FromBase64String(ivB64);
        var sealedBytes = Convert.FromBase64String(ctB64);

        // 2. Decrypt locally with the recovered DEK.
        var cipherLen = sealedBytes.Length - 16;
        var cipherBytes = sealedBytes[..cipherLen];
        var tag = sealedBytes[cipherLen..];
        var plain = new byte[cipherLen];
        using var aes = new AesGcm(dek, 16);
        aes.Decrypt(iv, cipherBytes, tag, plain);
        return Encoding.UTF8.GetString(plain);
    }

    public async Task<int> RotateKekAsync()
    {
        await PostRawAsync($"/v1/transit/keys/{_kek}/rotate", new { });
        using var resp = await _http.GetAsync($"/v1/transit/keys/{_kek}");
        var json = JsonDocument.Parse(await resp.Content.ReadAsStringAsync());
        return json.RootElement.GetProperty("data").GetProperty("latest_version").GetInt32();
    }

    public async Task ShredKekAsync()
    {
        // Crypto-shredding: allow deletion, then delete the KEK so no DEK can be unwrapped again.
        await PostRawAsync($"/v1/transit/keys/{_kek}/config", new { deletion_allowed = true });
        await _http.DeleteAsync($"/v1/transit/keys/{_kek}");
    }

    private async Task PostRawAsync(string path, object body)
    {
        var content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");
        await _http.PostAsync(path, content); // ignore "already exists" / "already mounted"
    }
}

File: 4-envelope-encryption-and-kms/0-typescript/src/crypto/envelope.service.ts
import { Injectable } from "@nestjs/common"
import { randomBytes, createCipheriv, createDecipheriv } from "crypto"
import { TransitService } from "../transit/transit.service"

export interface EncryptedEnvelope {
    wrappedDek: string // Vault Transit "vault:vN:..." blob — the DEK encrypted by the KEK.
    iv: string // base64 12-byte GCM nonce.
    authTag: string // base64 16-byte GCM authentication tag.
    ciphertext: string // base64 AES-256-GCM ciphertext of the plaintext.
}

/**
 * EnvelopeService implements the two-tier envelope scheme:
 *   data ── encrypted by ──▶ DEK (per-record, AES-256-GCM, generated locally)
 *   DEK  ── wrapped by ────▶ KEK (lives only in Vault Transit)
 * Bulk data never touches Vault; only the tiny 32-byte DEK is wrapped/unwrapped.
 */
@Injectable()
export class EnvelopeService {
    /** Generate a fresh DEK, AES-GCM encrypt the data, then wrap the DEK with the KEK. */
    async encrypt(plaintext: string): Promise<EncryptedEnvelope> {
        // 1. Generate a brand-new 256-bit DEK for THIS record only.
        const dek = randomBytes(32)
        const iv = randomBytes(12)

        // 2. Encrypt the bulk data locally with the DEK — Vault never sees the plaintext.
        const cipher = createCipheriv("aes-256-gcm", dek, iv)
        const ciphertext = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()])
        const authTag = cipher.getAuthTag()

        // 3. Wrap the DEK with the KEK inside Vault, then throw the raw DEK away.
        const wrappedDek = await TransitService.wrapDek(dek)
        dek.fill(0)

        return {
            wrappedDek,
            iv: iv.toString("base64"),
            authTag: authTag.toString("base64"),
            ciphertext: ciphertext.toString("base64"),
        }
    }

    /** Unwrap the DEK via the KEK, then AES-GCM decrypt the bulk data. */
    async decrypt(env: EncryptedEnvelope): Promise<string> {
        // 1. Ask Vault to unwrap the DEK (works across KEK versions after rotation).
        const dek = await TransitService.unwrapDek(env.wrappedDek)

        // 2. Decrypt locally with the recovered DEK.
        const decipher = createDecipheriv("aes-256-gcm", dek, Buffer.from(env.iv, "base64"))
        decipher.setAuthTag(Buffer.from(env.authTag, "base64"))
        const plaintext = Buffer.concat([
            decipher.update(Buffer.from(env.ciphertext, "base64")),
            decipher.final(),
        ])
        dek.fill(0)
        return plaintext.toString("utf8")
    }
}
